### ISSUE TITLE
Consider whether to allow multiple sessions in browsers

### ISSUE THREAD

**Issue by jgraham** [w3c/webdriver-bidi#103]
With classic WebDriver the blocking nature of the commands means that it only makes sense to have one active session at a time. But with BiDi we can imagine multiple clients connecting to the browser simultaneously and each getting its own set of events. For example alongside a client running tests, a seperate client might be checking performance metrics for each page load. These functions are theoretically quite decoupled but when only a single session is permitted the clients at least have to cooperate to share a session id and internal state like the event subscriptions. In practice this usually means everything has to be part of the same client (as in the devtools model). But that monolithic design might not be necessary and might improve the tooling ecosystem when people are able to make small tools that do one thing rather than needing to integrate at the code level.

**Comment by foolip** [comment]
I've read through the notes and it's not clear to me if any next steps were agreed upon.

If I'm reading everything correctly, this is about whether it's possible to establish multiple BiDi WebSocket connections to a single browser instance. It's *not* about the number of browser instances that could exist at the same time, and also not strongly related to WebDriver classic since possibility of multiple BiDi sessions exist with just a single classic session, or indeed with [no such session](https://github.com/w3c/webdriver-bidi/issues/97).

Being able to connect multiple times is somewhat related to the ability reestablishing a lost connection, since the "oh here's a new connection but I've already had one before" situation happens for both.

The main question here to me seems to be session isolation. Presumably event subscription would be per session, but there would definitely not be total isolation, there are lots of things that could be done in one session that would necessarily have some side effects in the other. Is that the kind of implementation difficulty that might happen here, or is it something else entirely?

**Comment by jgraham** [comment]
Yes, so the session holds some state explicitly, like the set of subscribed events per context. So that wouldn't be shared between sessions. But of course at some level all sessions are operating on a giant pile of state in the form of the browser itself. So of course different sessions would be affected with things happening in the other session.

I think the key usecase here is something like @christian-bromann mentioned where you have one session that's e.g. driving tests, and another session that's monitoring those tests. The user running the tests might not even have access to the code which does the monitoring; certianly it doesn't make sense that e.g. Selenium should need to care about saucelabs-specific monitoring code. Without the possibility to have multiple sessions at the browser level you'd need some kind of middleware that could multiplex multiple sessions over a single bidi connection, keeping track of all the state affecting which sessions could see which events (and that might become difficult/impossible if we ever have a situation where you can configure to either get events with property A or the same events with property B, but never both), and provide unique command ids, and so on.

Implementation wise I think there are two concerns:
* Multiple sessions means you really need to keep per-session state in a specific object tied to the session and not e.g. assume the session is a singleton or that you can attach the state directly to DOM objects.
* It's unclear how one specifies to connect to an existing session rather than creating a new one, in the general case. For gecko with one WebSocket per browser instance that's simple; if you know the websocket port you can create a session. For something like Selenium Grid (or maybe Safari? @gsnedders) there can be multiple sessions per server instance all sharing the same server and routing based on session id, that's harder.

**Comment by sadym-chromium** [comment]
Maybe something like this would be useful: https://github.com/GoogleChromeLabs/chromium-bidi#field-channel

The purpose of the `channel` is to allow ChromeDriver app to tunnel client's BiDi requests/events and to send own infra commands as well without them being interfered.

**Comment by gsnedders** [comment]
> For something like Selenium Grid (or maybe Safari? @gsnedders) there can be multiple sessions per server instance all sharing the same server and routing based on session id, that's harder.

@patrickangle what do we do currently with WebDriver Classic? you can have multiple sessions connected to different devices, right?

**Comment by whimboo** [comment]
Please note that per WebDriver spec there is only a single active session:

https://w3c.github.io/webdriver/#sessions

```
A remote end has an associated list of active sessions, which is a list of all sessions that are currently started. A remote end that is not an intermediary node has at most one active session at a given time. 
```

As such this topic is for WebDriver BiDi only (at least for now).

**Comment by jgraham** [comment]
It all depends which point of view you look at.

WebDriver classic only supports a single session at the level of a browser. That's almost necessarily true because of the blocking commands and global state. But of course intermediate nodes can handle many sessions at once, each connecting to a different browser.

WebDriver BiDi optionally allows multiple sessions for a single browser so that e.g. your test runner and network logger could be seperate tools which can independently establish sessions without having to coordinate by sharing a session id. However they might both connect to the same WebSockets server (since each server can handle more than one connection).

It's not really clear to me which case requires multiplexing multiple ""subsessions" over a single session (which afaict is the capability offered by the "channel" field; apologies if I misunderstood that).

**Comment by whimboo** [comment]
If there are multiple sessions it's not clear yet if those should share object and element references or not. If they should not share any data, then multiplexing might make sense? Having a debugger on a different session as the automation might probably not be that useful given that the right elements cannot be observed.

**Comment by jgraham** [comment]
Do you have an example of a case where we want to share identifiers out of band so that they can be reused across different automation tools, but where the tools otherwise want to partition state? It seems to me that if you're at the level of sharing element identifiers you're probably tightly integrated enough to also share other state. But maybe there are counterexamples?

**Comment by sadym-chromium** [comment]
There are several arguable assumptions:
1. We want to support multiple clients connected to the same browser instance (test runner + network logger).
2. We want to keep "session" identifying the browser instance. Sessions map 1:1 browser instances.
3. It seems natural to share the state and IDs within 1 session.
4. We want to split subscriptions per clients within 1 session.

It can be done either by:
5. Split "sub-sessions" by connection. Each new connection to the existing session is a new sub-session.
6. Restricting amount of connections to 1, and separating event loops by come special key.

From the protocol level the option **5** seems the most natural. From the specification level it will end up with having some internal sub-session identifier.

And here come our implementation details: we have to keep only 1 connection from ChromeDriver to Browser, and to differentiate sub-sessions, we exposed that internal sub-session identifier as `channel` extension.

**Comment by sadym-chromium** [comment]
So my suggestion is:
1. Allow multiple connections to the same session.
1. Add `session.connect` command, which creates a sub-session and shares the session's state and IDs, but having separate event buffers and subscriptions.

**Comment by whimboo** [comment]
Note that for `session.connect` the client has to know the session id to prevent hi-jacking of an already existing session.

**Comment by sadym-chromium** [comment]
Would the problem be solved, if in BiDi we introduce a way to distinguish if the user wants to create a new session in some specific browser instance or in a new instance?

|| **BiDi** | **Classic HTTP** |
|-|-|-|
| New Browser + New Session | **_[coliision]_** WS connect to `/session` | `POST /session` |
| Existing Browser + New Session | **_[coliision]_** WS connect to `/session` | Not supported | 
| Existing Session | WS connection to `session/SESSION_ID` | `/session/SESSION_ID` endpoint |

This can be achieved, by changing the [`session.new`](https://w3c.github.io/webdriver-bidi/#command-session-new) command behaviour. Currently, if there is a session associated with the WS connection, `session.new` will fail. Instead, in that case the command can create a new session in the same browser instance and re-associate the WS connection to that newly created session.

**Comment by gsnedders** [comment]
So where have we got to? What's the status with current implementations?

It feels like we should be re-examining what's in the session module, because the session.subscribe/session.unsubscribe commands would need connection-specific equivalents to subscribe to events for just a single connection. Would the session commands then affect _every_ connection to that session?

**Comment by tapashi15-coder** [comment]
Hi folks

Thanks for all the detailed discussion in this thread!

I’m currently working on a setup where:

A Puppeteer-based AWS Lambda function launches a Firefox browser using BiDi protocol.

A Go-based Lambda extension attempts to connect to the same BiDi session to listen for network events and generate a HAR file.

This is similar to how we previously used CDP, where tools could connect to the same session via /json endpoints and capture events independently.

From the comments above (especially [this one](https://github.com/w3c/webdriver-bidi/issues/103#issuecomment-1080082699) by @jgraham), I understand that BiDi sessions are not shareable—meaning only the original client can receive the subscribed events, and there's no pubsub-style behavior like in CDP.

Just wanted to confirm:

Is this still the current limitation, or are there any updates/plans to support multiple listeners or some form of external session introspection?

If not, is the only recommended workaround to proxy the events via the original BiDi client (e.g., Puppeteer relays BiDi network events to the extension)?

Appreciate any pointers or updates. Thanks again!

**Comment by jgraham** [comment]
The spec does support multiple sessions (this is not yet implemented in Firefox). However the sessions have (some degree of) isolated state (I'd prefer more, but there are limits since of course the browser itself is a big mass of shared state).

For the use case above you'd create one session and use it to drive the browser, and a second session which would subscribe to the relevant network events and collect the information needed to generate a HAR file. The only external state that you'd need to share between the sessions is the WebSockets URL to connect to the browser (i.e. you wouldn't share a session id).

Is there any reason that doesn't work for your use case?
